runtimes/aleph-debian-12-python/init1.py

```diff
@@ -0,0 +1,606 @@
+    config = receive_config(client)
+    setup_system(config)
+
+    try:
+        app: Union[ASGIApplication, subprocess.Popen] = setup_code(
+            config.code, config.encoding, config.entrypoint, config.interface
+        )
+        client.send(msgpack.dumps({"success": True}))
+    except Exception as error:
+        client.send(
+            msgpack.dumps(
+                {
+                    "success": False,
+                    "error": str(error),
+                    "traceback": str(traceback.format_exc()),
+                }
+            )
+        )
+        logger.exception("Program could not be started")
+        raise
+
+    class ServerReference:
+        "Reference used to close the server from within `handle_instruction"
+        server: asyncio.AbstractServer
+
+    server_reference = ServerReference()
+
+    async def handle_instruction(reader, writer):
+        data = await reader.read(1000_1000)  # Max 1 Mo
+
+        logger.debug("Init received msg")
+        if logger.level <= logging.DEBUG:
+            data_to_print = f"{data[:500]}..." if len(data) > 500 else data
+            logger.debug(f"<<<\n\n{data_to_print}\n\n>>>")
+
+        try:
+            async for result in process_instruction(
+                instruction=data, interface=config.interface, application=app
+            ):
+                writer.write(result)
+                await writer.drain()
+
+                logger.debug("Instruction processed")
+        except ShutdownException:
+            logger.info("Initiating shutdown")
+            writer.write(b"STOPZ\n")
+            await writer.drain()
+            logger.debug("Shutdown confirmed to supervisor")
+            server_reference.server.close()
+            logger.debug("Supervisor socket server closed")
+        finally:
+            writer.close()
+
+    server = await asyncio.start_server(handle_instruction, sock=s)
+    server_reference.server = server
+
+    addr = server.sockets[0].getsockname()
+    print(f"Serving on {addr}")
+
+    try:
+        async with server:
+            await server.serve_forever()
+    except asyncio.CancelledError:
+        logger.debug("Server was properly cancelled")
+    finally:
+        logger.warning("System shutdown")
+        server.close()
+        logger.debug("Server closed")
+        umount_volumes(config.volumes)
+        logger.debug("User volumes unmounted")
+
+
+if __name__ == "__main__":
+    logging.basicConfig(level=logging.DEBUG)
+    asyncio.run(main())
+
+    logger.info("Unmounting system filesystems")
+    system("umount /dev/shm")
+    system("umount /dev/pts")
+    system("umount -a")
+
+    logger.info("Sending reboot syscall")
+    # Send reboot syscall, see man page
+    # https://man7.org/linux/man-pages/man2/reboot.2.html
+    libc = ctypes.CDLL(None)
+    libc.syscall(169, 0xFEE1DEAD, 672274793, 0x4321FEDC, None)
+    # The exit should not happen due to system halt.
+    sys.exit(0)
```

Comment:
Is it absolutely necessary to duplicate this file?

Reply:
Not absolutely, but I suspect that it will evolve over time while we may want to limit updates to the Debian 11 runtime to the minimum maintenance.